From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sat, 15 Feb 2020 17:42:13 +0200
Subject: [PATCH] Allow proxying proxies

Disabled by default, this option is available for few hackish things

diff --git a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
index 55d42b9775ec274b21568a86303f0a9e0bbde1f5..a60866930c8862cefe6ebd905083f565411a36dc 100644
--- a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
@@ -96,4 +96,6 @@ public interface FireflyProxyConfig {
     boolean isLogPacketDecodeErrors();
 
     boolean isHandleLegacyPing();
+
+    boolean isAllowProxyingProxies();
 }
diff --git a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
index 0d4608bac359f5e784c0474616821c296de920e2..144d8b4c3f4e93ab9fa917e6e9b4bc70701076d0 100644
--- a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
@@ -87,6 +87,9 @@ public class FireflyConfiguration extends WaterfallConfiguration {
     @Getter
     private boolean handleLegacyPing = true;
 
+    @Getter
+    private boolean allowProxyingProxies = false;
+
     @Override
     public void load() {
         super.load();
@@ -112,6 +115,7 @@ public class FireflyConfiguration extends WaterfallConfiguration {
         dropConnectionsSendingInvalidPackets = config.getBoolean("packets.drop_invalid_packet_senders", dropConnectionsSendingInvalidPackets);
         logPacketDecodeErrors = config.getBoolean("packets.log_packet_decode_errors", logPacketDecodeErrors);
         handleLegacyPing = config.getBoolean("packets.handle_legacy_ping", handleLegacyPing);
+        allowProxyingProxies = config.getBoolean("allow_proxying_proxies", allowProxyingProxies);
     }
 
     private TCPFastOpenMode setupTfo(int value) {
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index e397a85f4630aedf222f98eae2ab8a2b4b0839b6..40c31d998390655a68db652c47c9e4e61bd4e488 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -280,6 +280,7 @@ public class DownstreamBridge extends PacketHandler
             String serverBrand = DefinedPacket.readString( brand );
             brand.release();
 
+            if (!bungee.getConfig().isAllowProxyingProxies()) // Firefly - Allow proxying proxies when enabled
             Preconditions.checkState( !serverBrand.contains( bungee.getName() ), "Cannot connect proxy to itself!" );
 
             brand = ByteBufAllocator.DEFAULT.heapBuffer();
